Add support for acme http-01 challenge

peschmae opened the pull request:
This is my work on implementing the ACME http-01 challenge.
It makes use of the SFTP destination, to avoid writing duplicate code for handling sftp connections. I'm not sure if it might make more sense to introduce a new type of destination plugin, that shares code with the SFTP one, but is specific just for acme/http challenges, to avoid confusion for the users.

Right now it doesn't support renewal of certificates yet, it's just the barebones implementation to create a new certificate using the HTTP-01 challenge.

Feel free to provide feedback to this pull request, I'll continue working on this for the next few days.

Currently lacks tests, I'll try to implement these as well.
Fixes #3164
Fixes #3154

hosseinsh reviewed (commented):
good work Mathias, I was also planing to add the http-01 feature, but you got ahead of me. 
I like the idea of defining a destination, and I have already expanded S3 bucket to support this feature. 
With S3 bucket, one can realize the approach, where the server redirects to a known S3 bucket for the token validation.

peschmae commented on lemur/plugins/lemur_acme/plugin.py:
```diff
@@ -715,3 +719,138 @@ def create_authority(options):
+        if challenge is None:
+            raise Exception('HTTP-01 challenge was not offered by the CA server.')
+        else:
+            # Here we probably should create a pending certificate and make use of celery, but for now
```
You're right. I was already thinking about how I need to handle the renewal of a certificate, which will need its own task. But for the initial authorisation, it's way easier to just do it all at once, and ignore pending certificates and such

peschmae replied:
I need to remove that comment...
I was able to renew certificates created using the http challenge, using the reissue cli command, now i just need to verify that it works when auto renewing using the celery task

peschmae commented:
@hosseinsh in #3157 there is the issue of using EAB with the ACME protocol. And thinking about it, I realised that maybe it might be more useful, to have a single ACME IssuerPlugin where you can choose the validation method, instead of having two dedicated IssuerPlugins. What do you think, is the better approach?

peschmae commented on lemur/plugins/lemur_acme/plugin.py:
```diff
@@ -715,3 +723,162 @@ def create_authority(options):
+            for destination in destinations:
+                # we only want to use sftp destinations here
+                if destination.plugin_name == "sftp-destination":
+                    self.destination_list.append(destination.label)
```
@hosseinsh Do you know, if theres a dynamic way, to create the options list?
Right now, when you add a new destination, it's not displayed automatically, and you need to restart lemur to initialise the list again :/

peschmae replied:
I've started refactoring this, so it's dynamically loaded using angular.

Similar to how the dns providers are done on the certificates screen, if a acme issuer is chosen

peschmae commented:
Fixes #3154

hosseinsh commented on lemur/plugins/lemur_acme/plugin.py:
```diff
@@ -715,3 +746,206 @@ def create_authority(options):
+
+        current_app.logger.info("Uploaded HTTP-01 challenge tokens, trying to poll and finalize the order")
+
+        # It is possible to set a deadline time.
```
shall we set a deadline? it defaults to 90s anyways.

hosseinsh replied:
for the DNS path we are currently setting it to 6min:

https://github.com/Netflix/lemur/blob/983f9beacb85cb1ecb8ce10a5028d2664dffed23/lemur/plugins/lemur_acme/plugin.py#L179

peschmae replied:
Right now I've set the timeout to 1minute, since we already handled the token upload. And unless there's an async `upload_acme_token` implementation (e.g. deploying a rule on a loadbalancer, that needs to be synced) we should be safe.

But I'll set it to the default 90s, and remove the comment. If there ever is an async upload implementation, it might be useful to make this configurable.

hosseinsh reviewed (commented):
@peschmae planing to dedicate more time to this PR the coming days.

With respect to where the http01 validation logic should reside, I was initially planing to use the same IssuerPlugin, and define the validation method within the request (e.g., as an issuer option). Basically in `def create_certificate(self, csr, issuer_options):` branch off to the http01 validation, if available.

However, I liked your idea of having two separate plugins, specially as it improves the plugin's readability. The DNS validation segment requires some updates and refactoring in the future, to become stateful and hence more performant. This separation could have made that improvement potentially easier.
  
However, the arguments from `the issue of using EAB with the ACME protocol` would warrant a single plugin, where via issuer options we would be able to define: DNS, HTTP-01, None, or other options in the future.
We can support backward compatibility, but defaulting to DNS if no validation method is provided in the options. 

This would also allow for less redundant code in the two plugins.

to summarize, I think one plugin would have more benefits, considering all the different requirements.

peschmae commented:
I've now refactored the way, how the destinations are displayed on the authority create screen.

It's now reloading them using the `DestinationService`, and I've introduced a new type of element `destinationSelect` so it can be reused, if there's need for it.

peschmae commented:
> However, I liked your idea of having two separate plugins, specially as it improves the plugin's readability. The DNS validation segment requires some updates and refactoring in the future, to become stateful and hence more performant. This separation could have made that improvement potentially easier.

That was also my idea, to avoid mixing the code base, but due to the AcmeHandler a lot of code could already be reused, and it's possible to refine that a bit more.
But if the DNS validation should be refactored anyway, I might leave it like this for now.

> However, the arguments from `the issue of using EAB with the ACME protocol` would warrant a single plugin, where via issuer options we would be able to define: DNS, HTTP-01, None, or other options in the future.
> We can support backward compatibility, but defaulting to DNS if no validation method is provided in the options.
I thought about that quite a bit, and unless we also change the frontend to allow for more dynamic options (at least for the ACME authorities), this might become quite overloaded, since the different options, require different fields.

Even when splitting between EAB and non-EAB providers (instead of splitting between challenge types), you end up with a few unnecessary fields if you don't display them dynamically.

No matter which of these two options is chosen, I think moving the `create_certificate` methods of `ACMEIssuerPlugin` and `ACMEHttpIssuerPlugin` into a (for example) `AcmeHttpChallenge` and `AcmeDnsChallenge` class, would be more optimal, and allow for a abstract `AcmeChallenge` class, which implement revoke_certificate for all of them and still allows for readable code while also enabling implementing further Challenge types, if they become relevant.

This way you could also easily switch between having separate IssuerPlugins, or just a dropdown in the `issuer_options`

hosseinsh commented:
> No matter which of these two options is chosen, I think moving the create_certificate methods of ACMEIssuerPlugin and ACMEHttpIssuerPlugin into a (for example) AcmeHttpChallenge and AcmeDnsChallenge class, would be more optimal, and allow for a abstract AcmeChallenge class, which implement revoke_certificate for all of them and still allows for readable code while also enabling implementing further Challenge types, if they become relevant.

+1 on this. 


Today, I learned about an important edge case that the current two plugin design doesn't support, namely, a SAN certificate request that requires mixing the validation methods. For instance, a wildcard SAN which would require DNS validation.
I am still figuring out if ACME,  this case Let's Encrypt, supports this use-case ([url](https://community.letsencrypt.org/t/mixed-domain-validation-types-http01-dns/135600)).

My initial intuition is this should be supported since, for each domain, all types of validations are returned in the Authorization object.

It certainly gets more tricky with how to decide on the validation method per domain. I need to do more thinking about this, but one idea could be to differentiate based on the presence of wildcard domain. For instance, in a plugin design, we would branch off to http-01 validation, unless a wildcard domain is present. In this case, a hybrid approach is required, during which we first `answer_challenge` for all non-wild-card SANs, and then trigger the normal DNS flow for the remaining DNS authorization.

This is not something we should address within this PR, but I wonder if this use-case would make the case for the one plugin design?

peschmae commented:
>Today, I learned about an important edge case that the current two plugin design doesn't support, namely, a SAN certificate request that requires mixing the validation methods. For instance, a wildcard SAN which would require DNS validation.

Wow, I haven't even thought about that. But according to your post in the community, it really does work,  but it seems to be so uncommon that not even certbot properly supports it but acme.sh does ([link](https://github.com/acmesh-official/acme.sh/wiki/How-to-issue-a-cert#3-multiple-domains-san-mode--hybrid-mode))

But I think this exceeds the scope of this pull request, and will create a separate ticket for it.

For now I will start refactoring the challenges into their own class, but leave it as two plugins, and would propose to merge it into a single authority (using a more dynamic UI) in the follow up ticket.

hosseinsh commented:
> But I think this exceeds the scope of this pull request, and will create a separate ticket for it.

agree. 

> For now I will start refactoring the challenges into their own class, but leave it as two plugins, and would propose to merge it into a single authority (using a more dynamic UI) in the follow up ticket.

Awesome. Thank you.

hosseinsh commented:
once Lemur's ACME plugin becomes stateful, such that the domain validation status is persisted (30 days valid for LE), it would be way more convenient to mix the validation types.

hosseinsh commented:
for completeness, adding also a reference to https://github.com/Netflix/lemur/pull/3192  for 
Expanding the S3 plugin with acme-http01-token support

peschmae commented:
Took me some time, to get back to this pull request.

But now I've refactored most of the challenge related code into `AcmeChallenge` classes, but also had to split the handlers from the plugin to avoid a circular import.

There are still some things I would like to optimise as well as improve the tests and rebase to master again

hosseinsh commented:
nice work Mathias! 
I gotta check out the latest changes for local testing and code review.

Keep up the good work :raised_hands::skin-tone-4:

peschmae commented:
I broke quite a few tests, with the refactoring, and also the previous few commits. But especially the refactoring, since I now have to mock for example 'current_app' multiple times within one test case.

I'm currently working on getting the tests to work again, but until now, I only needed to change which classes are patched, and no logic within the test, and a quick manual test, was successful, so I'm confident I didn't break too much. But i still need to invest some time to fix it all

hosseinsh commented:
yeah, a rebase first might be worth it.

hosseinsh commented:
Hi @peschmae, that is a milestone! nice work!
I haven't had the time to make progress on the http01 work as much as I was hoping, and needed to prioritize work on other corners of Lemur. Planing to spend time in the comings days to view and test this PR, and get it merged.

peschmae commented:
I'll make sure to merge master (with the changes from #3221) today or on the weekend

hosseinsh reviewed (commented):
@peschmae this has become an impressive PR. 
yesterday, I finally managed to perform some local tests, to test issuance of both DNS01 and HTTP01 cert issuance. We are using S3 for hosting the tokens, and it is working excellent. 

I noticed some improvements we can do to the way we handle exceptions during finalize order. Other than that the PR looks awesome. 

Thank you again!

PS: updating to the master branch would be excellent, as we have fixed some deprecation warnings and other issues.
one of the improvement features in the master branch is the way flask app is constructed for acme test, to avoid warnings.

hosseinsh commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+            ),
+        ).decode()
+
+        if current_app.config.get("IDENTRUST_CROSS_SIGNED_LE_ICA", False) \
```
one of the improvement features in the master branch is the way flask app is constructed for acme test, to avoid warnings.

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+
+        return response
+
+    def cleanup(self, challenge, acme_client, validation_target):
```
Implementing a cleanup makes sense, I'll add that as well. But it will need a delete_acme_token function within the S3Destination to make it work with the S3 plugin

hosseinsh replied:
good point, will take care of that in a separate PR.

hosseinsh replied:
raised the PR https://github.com/Netflix/lemur/pull/3245

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+
+        current_app.logger.info("Uploaded HTTP-01 challenge tokens, trying to poll and finalize the order")
+
+        finalized_orderr = acme_client.finalize_order(orderr, datetime.datetime.now() + datetime.timedelta(seconds=90))
```
I assumed, that the file upload is done within the 90 second timeout defined in `finalize_order`, and because right now, the cert bootstrap isn't done asynchronously I didn't add much error handling.

But it would definitely be better to add a nice and readable error message that is displayed to the user, instead of the ACME errors

peschmae replied:
I've added quite a bit more exception handling to the http challenge, as well as the upload function of the sftp destination

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+        authority = issuer_options.get("authority")
+        acme_client, registration = self.acme.setup_acme_client(authority)
+
+        orderr = acme_client.new_order(csr)
```
`orderr.uri` is already logged by the acme_client in my setup.
Would you wan't to log it with the Exception if no challenges are offered, or where did you intend to add it?

hosseinsh replied:
I think the exception was missing that info, and it would add value to for debugging

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+                    chall.append(i)
+
+        if len(chall) == 0:
+            raise Exception('HTTP-01 challenge was not offered by the CA server.')
```
I implemented a check, if the authorization is already in a valid state and skip it if that's the case.

If no challenges are in an invalid state, I now proceed directly to `poll_and_finalize`

hosseinsh replied:
that is an excellent approach.

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,240 @@
+
+        response, validation = challenge.response_and_validation(acme_client.net.key)
+
+        destination_plugin.upload_acme_token(challenge.chall.path, validation, destination.options)
```
I've added exception handling in the create_certificate function, so if the upload fails, it's displayed to the user, and more information logged to the log

peschmae commented on lemur/plugins/lemur_acme/plugin.py:
```diff
@@ -779,3 +269,144 @@ def create_authority(options):
+            raise InvalidConfiguration("There is no ACME account saved, unable to revoke the certificate.")
+        acme_client, _ = self.acme.setup_acme_client(certificate.authority)
+
+        fullchain_com = jose.ComparableX509(
```
it's now a function in the acme_handlers

hosseinsh commented:
Thanks for the quick workaround @peschmae, and addressing all the feedback.

charhate commented on lemur/plugins/lemur_acme/acme_handlers.py:
```diff
@@ -0,0 +1,516 @@
+            result = dns.resolver.query(domain, 'CNAME')
+            if len(result) > 0:
+                return str(result[0].target).rstrip('.')
+        except dns.exception.DNSException:
```
Should this exception be captured in Sentry or log?

hosseinsh commented on lemur/plugins/lemur_sftp/plugin.py:
```diff
@@ -150,39 +184,119 @@ def upload(self, name, body, private_key, cert_chain, options, **kwargs):
-                current_app.logger.debug("Creating {0}".format(dst_path_cn))
-                sftp.mkdir(dst_path_cn)
-            except IOError:
+            # upload certificate files to the sftp destination
```
type: upload --> delete

hosseinsh commented on lemur/plugins/lemur_sftp/plugin.py:
```diff
@@ -121,6 +118,43 @@ def upload(self, name, body, private_key, cert_chain, options, **kwargs):
+        ssh_priv_key = self.get_option("privateKeyPath", options)
+        ssh_priv_key_pass = self.get_option("privateKeyPass", options)
+
         # upload files
```
upload --> delete

hosseinsh reviewed (approved):
This PR seems to be in a good state to be merged. Local testing is looking promising so far. 

There are some minor open issues, which we can address within this PR, time permitting, or future PRs.

Planing to merge this PR tomorrow.

peschmae commented on lemur/plugins/lemur_sftp/plugin.py:
```diff
@@ -121,6 +118,43 @@ def upload(self, name, body, private_key, cert_chain, options, **kwargs):
+        self.upload_file(dst_path, files, options)
+
+    # this is called from the acme http challenge
+    def delete_acme_token(self, token_path, token, options, **kwargs):
```
done

peschmae commented on lemur/plugins/lemur_acme/challenge_types.py:
```diff
@@ -0,0 +1,273 @@
+
+        return response, validation
+
+    def cleanup(self, token_path, token, validation_target):
```
Yeah, that was a bit too lazy of a copy&paste from the upload function.

But i've now removed the token from the cleanup, and delete_acme_token function

peschmae commented:
@hosseinsh I've implemented the remaining feedback (like removing the token value), and while at it, added tests for the `lemur_sftp` plugin.

I also found some more duplicate code, that I cleaned up. And now it's really ready for the merge.

hosseinsh commented:
nice work with deduplicating the cert extraction, it makes things easier for any changes on pinning to the cross-signed ICA.
all changes look good to me. 

Merging now.

Thanks again @peschmae for the PR. It was fun working wit you on this PR.
